test/distributed/test_c10d.py

```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
             x = checkpoint(self.l2, x)
             return x
 
-    def _test_ddp_checkpointing(self, checkpoint_once, process_group, use_bucket_view, find_unused_parameters=False):
+    def _prepare_dummy_data(self):
+        ddp_bs = 16
+        bs = ddp_bs * self.world_size
+        input = torch.rand((bs, 20), device="cuda", requires_grad=True)
+        target = torch.randn((bs, 20), device="cuda")
+        offset = self.rank * ddp_bs
+        ddp_input = input[offset: offset + ddp_bs]
+        ddp_target = target[offset: offset + ddp_bs]
+        return input, ddp_input, target, ddp_target
+
+    def _train_model(self, model, input_var, target, loss, run_checkpoint=False):
+        model.train()
+        if run_checkpoint:
+            output = checkpoint(model, input_var)
+        else:
+            output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+
+    def _test_ddp_checkpointing(
+        self,
+        input_model,
+        process_group,
+        use_bucket_view,
+        find_unused_parameters=False,
+        static_graph=False,
+        run_checkpoint=False
+    ):
         # to reprodce the same training results
         torch.cuda.set_device(self.rank)
         torch.manual_seed(31415)
-        if checkpoint_once:
-            model = self.CheckpointOnceModule().cuda()
-        else:
-            model = self.CheckpointTwiceModule().cuda()
-        model = nn.parallel.DistributedDataParallel(model,
-                                                    bucket_cap_mb=1,
-                                                    gradient_as_bucket_view=use_bucket_view,
-                                                    device_ids=[self.rank],
-                                                    process_group=process_group,
-                                                    find_unused_parameters=find_unused_parameters)
-        input_tensor = torch.rand((64, 2000), device="cuda", requires_grad=True)
-        output_tensor = model(input_tensor)
-        output_tensor.sum().backward()
-        return model
+        model = copy.deepcopy(input_model).cuda()
+        ddp_model = copy.deepcopy(input_model).cuda()
+        ddp_model = nn.parallel.DistributedDataParallel(
+            ddp_model,
+            bucket_cap_mb=1,
+            gradient_as_bucket_view=use_bucket_view,
+            device_ids=[self.rank],
+            process_group=process_group,
+            find_unused_parameters=find_unused_parameters
+        )
+        if static_graph:
+            ddp_model._set_static_graph()
+        self.assertEqual(ddp_model.get_ddp_logging_data().static_graph, static_graph)
+        input, ddp_input, target, ddp_target = self._prepare_dummy_data()
+        loss = nn.MSELoss()
+        for i in range(5):
+            model.zero_grad(set_to_none=False)
```

Comment:
What's default arg for `zero_grad`? Do users need to explicitly add `set_to_none=False` if they are using static graph training?

Reply:
in default, set_to_none=False, we do not have to set it explicitly